Add caching allocator interface

pxl-th opened the pull request:
Since Julia's GC is not aware of GPU memory, in scenarios with lots of allocations we end up in either OOM situations or in excessively high memory usage.
Even though the program may require only fraction of it.

To help with GPU memory utilizaton in a program with repeating blocks of code, we can wrap those regions in a scope that will utilize caching allocator every time the program enters this scope during the execution.

For example, this is especially useful when training models, where you compute loss, gradients w.r.t. loss and perform in-place parameter update of the model.

```julia
cache = GPUArrays.AllocCache()
model = ...
for i in 1:1000
    GPUArrays.@enable cache begin
        loss, grads = ...
        update!(optimizer, model, grads)
    end
end
```

Caching is done on: `(ArrayType, current device, eltype, dims[, buffer type])`.

### Example

In the following example we apply caching allocator at every iteration of the for-loop.
Every iteration requires 8 GiB of gpu memory, without caching allocator
GC wouldn't be able to free arrays in time resulting in higher memory usage.
With caching allocator, memory usage stays at exactly 8 GiB.

After the loop, we free all cached memory if there's any.
Alternatively, it will be freed automatically when cache is collected by GC.

```julia
cache = GPUArrays.AllocCache(CuArray)
n = 1024^3
CUDA.@sync for i in 1:1000
    GPUArrays.@enable cache begin
        sin.(CUDA.rand(Float32, n))
    end
end
GPUArrays.unsafe_free!(cache)
```

### Performance impact

Executing [GaussianSplatting.jl](https://github.com/JuliaNeuralGraphics/GaussianSplatting.jl/pull/26) benchmark (1k training iterations) on RX 7900XTX:

||Without caching allocator|With caching allocator|
|-|-|-|
|GPU memory utilization|![image](https://github.com/user-attachments/assets/92e0e802-9784-4e5a-85d0-7dfa7b4e8dbf)|![image](https://github.com/user-attachments/assets/ab4c68eb-2caa-4a56-9c0a-9176285ef66d)|
|Time|`59.656476` seconds|`46.365646` seconds|

### TODO

- [x] Support for 1.10.
- [x] Support bulk-freeing instead of caching.
- [x] Add PR description.
- [x] Documentation.
- [x] Tests.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,155 @@
+    # Otherwise, try fetching from `free`.
+    while !isempty(free_pool) && x ≡ nothing
+        tmp = pop!(free_pool)
+        # Array was manually freed via `unsafe_free!`.
+        storage(tmp).freed && continue
```
Not necessarily only by calling `unsafe_free!`, but also by the GC, right?

pxl-th replied:
In this particular place only by `unsafe_free!`, because cached arrays are stored either in `free` or in `busy` arrays, thus preventing GC from collecting them.

maleadt reviewed (commented):
Could you add some high-level design description to the PR?

As I mentioned on Slack, CUDA already has a caching allocator, so I'm not sure if for those back-ends this shouldn't boil down to basically batch-calling `unsafe_free!` at the end of each iteration, instead of actively caching arrays. Would be good to compare performance, if possible.

pxl-th commented:
@maleadt, I've updated the PR.
Also, I've added tests but they are not enabled right now, because no backend currently has the implementation merged (including JLArrays, because tests use released version of it).

Let me know what you think.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,251 @@
+"""
+cache_allocator(::Backend) = error("Not implemented.")
+
+"""
+    device(::Backend)
+
+Given KernelAbstractions `backend`, return current device.
+Each GPU backend must implement this.
+"""
+device(::Backend) = error("Not implemented.")
```
This seems like something that should go in KernelAbstractions.jl itself?

Alternatively, when keying on the array type, I think I could live with a `device(::AbstractGPUArray)` interface method.

pxl-th replied:
I've made it rely on array type instead of KA backend.

maleadt commented on docs/src/interface.md:
````diff
@@ -23,10 +23,18 @@ end
+## Caching Allocator
+
+```@docs
+GPUArrays.@cache_scope
+GPUArrays.@no_cache_scope
+GPUArrays.invalidate_cache_allocator!
````
Some bikeshedding ideas:

```suggestion
GPUArrays.AllocCache.@cached
GPUArrays.AllocCache.@uncached
# or @enable / @disable?

GPUArrays.AllocCache.invalidate!
```

IMO `@cache_scope` is pretty vague, it's not clear what's being cached, and a macro is always scoped anyway.

pxl-th replied:
Agree, changed to `@enable`, `@disable`, `invalidate!`.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,251 @@
+Given KernelAbstractions `backend`, return corresponding `PerDeviceCacheAllocator` for it.
+Each GPU backend must implement this.
+"""
+cache_allocator(::Backend) = error("Not implemented.")
```
Instead of passing in a KA back-end, what about keying everything on the GPU array type? It seems more natural, as the KA back-end doesn't seem to be used anyway.

pxl-th replied:
The reason for passing backend is because I was using it constantly when writing backend-agnostic code as this is something that allows KA kernel compilation (`kernel(kab)(args...)`) and transfer to GPU/CPU (`adapt(kab, host_array)`).
So this was my main "anchor" around backend-agnostic code.

Maybe we should also add an interface to obtain KA backend from the array type
`KA.get_backend(::Type{AMDGPU.ROCArray}) = ROCBackend()`
Because currently we only have an interface for obtaining backend from the actual array
`KA.get_backend(::AMDGPU.ROCArray) = ROCBackend()`

Which is cumbersome and leads to a situations where you have to create 0-sized arrays to obtain KA backend.

> Alternatively, when keying on the array type, I think I could live with a device(::AbstractGPUArray) interface method.

Same here, I'd rather query on `::Type{AbstractGPUArray}`, not on the actual array.

maleadt replied:
> Maybe we should also add an interface to obtain KA backend from the array type

That seems reasonable to me, cc @vchuravy. It's something that exists in Base as well for functions like `eltype`.

On the other hand, it may not be possible when the back-end encodes specific information like the device being used, but I don't know if that was ever the intent.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,251 @@
+This is useful to apply in a repeating block of code to avoid relying on
+GC to free gpu memory in time.
+
+`name` is a `Symbol` that defines which allocator to use
+(`:none` is reserved and means no allocator).
```
Wouldn't `nothing` be more idiomatic?

pxl-th replied:
Done.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,251 @@
+If no such allocation is found, execute `alloc_f` that does actual allocation,
+store it in cache for future use and return it.
+
+`skip_free::Bool` is used together with `PerDeviceCacheAllocator.free_immediately`.
+When `true` arrays are bulk-freed instead of stored in cache.
+In this case `alloc!` will avoid looking into "free" part of `cache`
+and execute `alloc_f` immediately, storing allocation for future bulk-freeing.
```
`skip_free` seems like an implementation detail?

The `free_immediately` flag also feels like a hack; have you benchmarked the difference between relying on CUDA's allocator vs. fully caching in GPUArrays? I'd be happy to not support this at all (i.e., have CUDA.jl also cache allocations in GPUArrays.jl instead of relying on the CUDA caching allocator) to simplify the API, especially if it would make sense performance-wise.

pxl-th replied:
The difference is pretty minor, around 2% improvement with `free_immediately=false`.

maleadt replied:
Yeah, let's drop it. That way it also works for people who run with the caching allocator disabled (i.e., HPC folks relying on UCX/MPI).

pxl-th replied:
Done. I've also uncommented tests so that JLArrays runs them. But other backends will fail for now on them.

pxl-th commented:
One difference I've found between Julia 1.10 and Julia 1.11:

- Julia 1.10:

```julia

julia> GPUArrays.AllocCache.@enable CuArray :loop begin
           x1 = CuArray(rand(Float32, 1))
       end
1-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 0.680597

julia> x1
ERROR: UndefVarError: `x1` not defined
```

- Julia 1.11:

```julia
julia> GPUArrays.AllocCache.@enable CuArray :loop begin
           x1 = CuArray(rand(Float32, 1))
       end
1-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 0.7703809

julia> x1
1-element CuArray{Float32, 1, CUDA.DeviceMemory}:
 0.7703809
```

Not sure where is it coming from.

maleadt commented on lib/JLArrays/src/JLArrays.jl:
```diff
@@ -387,4 +395,12 @@ Adapt.adapt_storage(::JLBackend, a::Array) = Adapt.adapt(JLArrays.JLArray, a)
 
+# Caching Allocator.
+
+const JLACacheAllocator = GPUArrays.AllocCache.PerDeviceCacheAllocator(JLArray)
+
+GPUArrays.AllocCache.cache_allocator(::Type{<:JLArray}) = JLACacheAllocator
```
Why is this needed, now that you switched to the array type? Isn't all information there for the caller to construct an appropriate allocator cache?

pxl-th replied:
This is for the internal implementation to retrieve the actual cache for `@enable`.
E.g. when CUDA calls [`alloc!`](https://github.com/JuliaGPU/GPUArrays.jl/pull/576/files#diff-fcdd73df7e8055bf03e5d08337fcd21346ddb4d4f4f6de2cf82e9979d1e1df7fR103) we retrieve its allocator cache based on its array type.
Otherwise the user would have to pass the cache itself to the macro, no?

maleadt replied:
I mean that we can get rid of the alias, and replace calls to `cache_allocator` by `AllocCache.PerDeviceCacheAllocator(AT)`. Just trying to minimize the interface to be implemented by back-ends.

pxl-th replied:
I'm not opposed, just a bit confused.
`AllocCache.PerDeviceCacheAllocator(AT)` is call to ctor, when with `cache_allocator` we retrieve an instance.
In CUDA.jl we [define a global variable](https://github.com/JuliaGPU/CUDA.jl/pull/2593/files#diff-1177a809b2ffc09e77e53c0c8a036ca503494f88c39c10b6027542670d444b0aR1) that we then retrieve with `cache_allocator`.
How are we going to access this variable with `AllocCache.PerDeviceCacheAllocator(AT)` if its a call to ctor?

Or is the point just to rename `cache_allocator` to `AllocCache.PerDeviceCacheAllocator`? But then it is ambiguous because ctor has the same method signature.

maleadt commented on src/host/allocations_cache.jl:
```diff
@@ -0,0 +1,240 @@
+    return sz
+end
+
+"""
+    invalidate!(AT::Type{AbstractGPUArray}, name::Symbol)
+
+Free all memory held by `name`d cached allocator given array type `AT`.
+"""
+invalidate!(AT::Type{<:AbstractGPUArray}, name::Symbol) =
+    invalidate!(cache_allocator(AT), device(AT), name)
```
Is it expected for users to need this? Why not have them wrap code in multiple `@enable` blocks?

pxl-th replied:
Yes, because the cache caches arrays based on their dims, there may be a situation where dims change (e.g. different batch size or the number of parameters of the model change) you need to invalidate the cache, because with new dims the old ones won't be retrieved.

E.g. with GaussianSplatting, where I [enable cache](https://github.com/JuliaNeuralGraphics/GaussianSplatting.jl/blob/79a0076aee5fca46a0ac7f3991976169fe4d82f2/src/training.jl#L186) for the training step.
But at some point the number of parameters of the model changes so we need to [invalidate the cache](https://github.com/JuliaNeuralGraphics/GaussianSplatting.jl/blob/79a0076aee5fca46a0ac7f3991976169fe4d82f2/src/training.jl#L224), because old dims are not used anymore.

> Why not have them wrap code in multiple @enable blocks?

IIUC, you mean something like this?
```julia
GPUArrays.AllocCache.@enable CuArray :train_step begin
    # some code
end

# some code outside of caching

GPUArrays.AllocCache.@enable CuArray :train_step begin
    # some code
end
```

But then again when you either no longer need the cache (done training) or the dims change you need to somehow invalidate it.

maleadt replied:
Oh, so the cache persists until the user calls `invalidate!`? I somehow missed that. It seems like a dangerous design to me; if you forget to `invalidate!` on any path outside of the `@enable`, memory will leak?

pxl-th replied:
Yes, although I think this is fine tradeoff. As a last resort we could invalidate the cache in alloc/retry mechanism, registering a hook. Similar to how we do with fft handle cache

pxl-th commented:
`ScopedValues.jl` on Julia 1.10 introduce a scope:

```julia
julia> using ScopedValues

julia> x = ScopedValue(1)
ScopedValue{Int64}(1)

julia> @with x => 2 begin
           x2 = x[]
           x3 = 1
       end
1

julia> x2
ERROR: UndefVarError: `x2` not defined
```

maleadt commented:
Another fundamental question (sorry for stretching this out): Why do you even care about the array type in the `@enable` interface? Wouldn't it be better if the user didn't have to worry about this? The cache can be sharded internally so that back-ends can `invalidate!` only their portion (e.g. from within a `retry_reclaim` hook), but I don't see why it _has_ to be provided.

Maybe the cache name should be optional as well. It could default to something derived from the current task's name, so that's it's really convenient to do:

```julia
AllocCache.@enable begin
  for i in epocs
    ...
  end
end
AllocCache.invalidate!()
```

Just spitballing here, you probably have a better view regarding it based on your experiments with it already.

---

Seeing the above written out, I wonder if a wholly different API wouldn't be much more idiomatic, reifing the now implicit stuff like the name of the cache:

```julia
cache = AllocCache()
cache() do
  for i in epocs
    ...
  end
end
empty!(cache)
```

A finalizer could then also empty the cache, avoiding the risk of leaking memory if you forget to `empty!` it.

pxl-th commented:
In the second approach, how are we going to pass it to the ctor of an array to cache without having some global structure to hold it without breaking much of the code?

And I'd be in favor of a macro that would avoid closure creation.

pxl-th commented:
> Seeing the above written out, I wonder if a wholly different API wouldn't be much more idiomatic, reifing the now implicit stuff like the name of the cache:

@maleadt , I've updated the implementation based on this, see examples in PR description for TL;DR.
It will now also free all memory in the finalizer.

maleadt commented on src/host/alloc_cache.jl:
```diff
@@ -0,0 +1,152 @@
+
+See [`@disable`](@ref).
+"""
+macro enable(cache, expr)
+    return quote
+        res = @with $(esc(ALLOC_CACHE)) => $(esc(cache)) $(esc(expr))
+        free_busy!($(esc(cache)))
+        res
+    end
+end
+
+"""
+    disable(expr)
+
+Evaluate expression `expr` without using allocations cache.
+This is useful to call from within `@enable` to avoid caching some allocations.
+"""
+macro disable(expr)
+    return quote
+        @with $(esc(ALLOC_CACHE)) => nothing $(esc(expr))
+    end
+end
```
Seeing how these are in the main GPUArrays module, I'd opt for more descriptive names. `@cached` and `@uncached` maybe?

Alternatively, with do-block syntaxes always taking the `cache` argument the intent would be clear (`enable(cache) do ... end`, `disable(cache) do ... end`), but I take it you're purposefully avoiding closures so that you don't run into boxing issues?

pxl-th replied:
Renamed.

> but I take it you're purposefully avoiding closures so that you don't run into boxing issues?

With macro I like that you can define variables and don't need to return them as would be with the do-block syntax at the end.

maleadt commented:
Looks good! Thanks for keeping up with my review requests.

CI failures look related though.

pxl-th commented:
CI failures will be fixed when we tag a new version of ScopedValues.jl that fixes introduction of a scope: https://github.com/vchuravy/ScopedValues.jl/pull/21#event-15858481695

maleadt commented:
> CI failures will be fixed when we tag a new version of ScopedValues.jl that fixes introduction of a scope: [vchuravy/ScopedValues.jl#21 (comment)](https://github.com/vchuravy/ScopedValues.jl/pull/21#event-15858481695)

Has been tagged already? https://github.com/vchuravy/ScopedValues.jl/releases/tag/v1.3.0

pxl-th commented:
Yep

pxl-th commented:
And the failures for CUDA/Metal/OneAPI are because those backends don't yet support the functionality

maleadt commented:
Pushed a simplification to the back-end interface to avoid having to reach into the scoped value. The only cost is having to allocate the key tuple unconditionally, but I think that should be fine.

pxl-th commented:
Looks good!

maleadt commented:
The `show` method is broken:

```julia
julia> cache = GPUArrays.AllocCache(CuArray)
GPUArrays.AllocCache{CuArray}(Error showing value of type GPUArrays.AllocCache{CuArray}:
ERROR: TypeError: in typeassert, expected Int64, got a value of type UInt64
```

Would probably be useful if it showed some stats.

maleadt commented on src/host/alloc_cache.jl:
```diff
@@ -0,0 +1,157 @@
+    using Base.ScopedValues
+end
+
+mutable struct AllocCache{T <: AbstractGPUArray}
+    lock::ReentrantLock
+    busy::Dict{UInt64, Vector{T}} # hash(key) => GPUArray[]
+    free::Dict{UInt64, Vector{T}}
```
Why lock down the type here? `T=CuArray` is going to result in insufficiently typed values anyway. Wouldn't it be better to make it a `Dict{UInt64, Any}` so that the user doesn't have to specify the array type?

pxl-th replied:
Good point. We are specifying the array type in CUDA/AMDGPU/etc ctor anyway

pxl-th commented:
> The `show` method is broken:
> 
> ```julia
> julia> cache = GPUArrays.AllocCache(CuArray)
> GPUArrays.AllocCache{CuArray}(Error showing value of type GPUArrays.AllocCache{CuArray}:
> ERROR: TypeError: in typeassert, expected Int64, got a value of type UInt64
> ```
> 
> Would probably be useful if it showed some stats.

Fixed.

maleadt commented:
Just saw this:

```
      From worker 2:	error in running finalizer: ErrorException("Invalidating allocations cache that's currently in use. Invalidating inside `@cached` is not allowed.")
      From worker 2:	error at ./error.jl:35
      From worker 2:	macro expansion at /home/tim/Julia/pkg/GPUArrays/src/host/alloc_cache.jl:75 [inlined]
      From worker 2:	macro expansion at ./lock.jl:273 [inlined]
      From worker 2:	unsafe_free! at /home/tim/Julia/pkg/GPUArrays/src/host/alloc_cache.jl:73
      From worker 2:	unknown function (ip: 0x7385147b5df2)
      From worker 2:	run_finalizer at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/gc.c:299
      From worker 2:	jl_gc_run_finalizers_in_list at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/gc.c:391
```

`@cached` should probably root the cache until it completes?

pxl-th commented:
Oh... right
